File: testClients/phelps/AI.py
#-*-python-*-
from BaseAI import BaseAI
from GameObject import *
from operator import itemgetter
import random
import math
import time

energyIndex = 0
carnIndex = 1
herbIndex = 2
defIndex = 3
speedIndex = 4

plantI = 1
CreatI = 2

class tileD(object):
  def __init__(self, x, y, dist):
    self.x = x
    self.y = y
    self.dist = dist

  def __repr__(self):
    return repr((self.x, self.y, self.dist))
    
class Cell(object):
  def __init__(self, x, y):
    self.x = x
    self.y = y

class AI(BaseAI):
  """The class implementing gameplay logic."""
  @staticmethod
  def username():
    return "phelps"

  @staticmethod
  def password():
    return "password"
    
  def distance(self,x1,y1,x2,y2):
    return math.sqrt((x2-x1)**2 + (y2-y1)**2)   
  
  def distancel(self, x, y):
    return math.sqrt((y[0]-x[0])**2 + (y[1]-x[1])**2)
  
  def distanceC(self, c1, c2):
    return math.sqrt((c2.x - c1.x)**2 + (c2.y - c1.y)**2)
  
  def init(self): 
    pass
  ##This function is called once, after your last turn
  def end(self):
    pass

  def getGridN(self, x):
    return self.grid[x[0]][x[1]]
  
  def getGrid(self, x, y):
    return self.grid[x][y]
  
  def neighborNodes(self, x, y):
    validNodes = []
    if x+1 < self.mapWidth:
      validNodes.append([x+1, y])
    if y+1 < self.mapHeight:
      validNodes.append([x, y+1])
    if x-1 >= 0:
      validNodes.append([x-1, y])
    if y-1 >= 0:
      validNodes.append([x, y-1])      
    return validNodes  
  
  def validMoves(self, x, y):
    moves = self.neighborNodes(x, y)
    
    for plant in self.plants:
      for cell in moves:
        if plant.x == cell[0] and plant.y == cell[1]:
          moves.remove(cell)
    
    for creature in self.creatures:
      for cell in moves:
        if creature.x == cell[0] and creature.y == cell[1]:
          moves.remove(cell)

    if len(moves) > 0:
      return moves
    else:
      return []
    
  def findDirection(self, startX, startY, goalX, goalY):
    
    s = [startX, startY]
    g = [goalX, goalY]
    
    pMoves = self.validMoves(s[0], s[1])    
    gMoves = [tileD(k[0], k[1], self.distance(k[0], k[1], g[0], g[1])) for k in pMoves]

    #for k in range(len(pMoves)):
    #  gMoves[0][k] = pMoves[k]
    #  gMoves[1][k] = (math.floor(self.distance(pMoves[k][0], pMoves[k][1], g[0], g[1])))  
      
#    dict = {cell:math.floor(self.distance(cell[0], cell[1], g[0], g[1])) for cell in pMoves}
       
   # bestMove = min(dict, key=dict.get)

    gMoves.sort(key=lambda tile: tile.dist)
    
    if len(gMoves) > 0:
      bestMove = [gMoves[0].x, gMoves[0].y]
      return bestMove
    else:
      return []
    
  def findAPathD(self, startX,startY, goalX, goalY, moveC):
    
    start = [startX, startY]
    goal = [goalX, goalY]
    
    closedSet = []
    openSet = [start]
    
    for k in range(moveC):      
      nextMove = self.findDirection(openSet[0][0], openSet[0][1], goal[0], goal[1])
      if len(nextMove) > 0:
        closedSet.append(nextMove)
        openSet.append(nextMove)
      else:
        return closedSet

      openSet.pop(0)

    return closedSet
  
  def findAPath(self, startX, startY, goalX, goalY):
    
    start = [startX, startY]
    goal = [goalX, goalY]
    
    closedSet = []
    openSet = [start]
    
    while len(openSet) > 0:
      if len(openSet) > 20:
        break
      nextMove = self.findDirection(openSet[0][0], openSet[0][1], goal[0], goal[1])
      closedSet.append(nextMove)
      if self.distance(openSet[0][0], openSet[0][1], goal[0], goal[1]) > 3:
        openSet.append(nextMove)
        
      openSet.pop(0)
    return closedSet
  
  def moveToMiddle(self, creature):
    x_mid = math.floor(self.mapWidth/2)
    y_mid = math.floor(self.mapHeight/2)
    
    path = self.findAPathD(creature.x, creature.y, x_mid, y_mid, creature.movementLeft)   
    while len(path) > 0 and creature.movementLeft > 0 and math.floor(self.distance(creature.x, creature.y, x_mid, y_mid)) > 2:
      creature.move(path[0][0], path[0][1])
      path.remove(path[0])
  
    return
  
  def findNearestFriendlyCreatureXY(self, creature):
    #Doesn't work if I only have one creature left
    myCreatures = [creature for creature in self.creatures if creature.owner == self.playerID and creature.currentHealth > 0]
    if len(myCreatures) > 1:    
      #Creates a dictionary of the distances to each creature I control that isn't at the calling creature's position
      dict = {target.id:math.floor(self.distance(creature.x, creature.y, target.x, target.y)) for target in myCreatures if target.x != creature.x and creature.y != target.y}    
      #Returns the object that is the closest creature   
      if len(dict) > 0:      
        bestCreature = next(target for target in myCreatures if target.id == min(dict, key=dict.get))
        #Returns the x and y value of that creature's location
        return bestCreature     
    return None
  
  def findNearestEnemyCreatureXY(self, creature):   
    #Creates a dictionary of the distances to each enemy creature
    enemyCreatures = [creature for creature in self.creatures if creature.owner != self.playerID and creature.currentHealth > 0]
    dict = {target.id:math.floor(self.distance(creature.x, creature.y, target.x, target.y)) for target in enemyCreatures if target.currentHealth > 0}  
    #Returns the object that is the closest creature      
    if len(dict) > 0:
      bestID = min(dict, key=dict.get)
      bestCreature = next(target for target in enemyCreatures if target.id == bestID)
      #Returns the x and y value of that creature's location
      return bestCreature 
    return None
  
  def findNearestFriendlyX(self, creature, X):
    myCreatures = [creature for creature in self.creatures if creature.owner == self.playerID and creature.currentHealth > 0]
    specializedCreatures = {"carnivorism":[], "herbivorism":[], "speed": []}
    
    for creature in myCreatures:
      stats = {"carnivorism": creature.carnivorism, "herbivorism" : creature.herbivorism}
      specializedCreatures[max(stats, key=stats.get)].append(creature)
    
    creatureList = specializedCreatures[X]
    
    dict = {target.id:math.floor(self.distance(creature.x, creature.y, target.x, target.y)) for target in creatureList if target.x != creature.x and creature.y != target.y}
    
    if len(dict) > 0:
      bestCreature = next(target for target in myCreatures if target.id == min(dict, key=dict.get))
      return bestCreature
    return None
  
  def findClosestXAndBreed(self, creature, bestStat):
    father = self.findNearestFriendlyX(creature, bestStat)
    if father is not None:
      path = self.findAPathD(creature.x, creature.y, father.x, father.y, creature.movementLeft)
      while len(path) > 0 and creature.movementLeft > 0 and self.distance(creature.x, creature.y, father.x, father.y) > 1 and creature.currentHealth > self.healthPerMove:
        creature.move(path[0][0], path[0][1])
        path.remove(path[0])
      if self.distance(creature.x, creature.y, father.x, father.y) == 1 and creature.canBreed == True and creature.currentHealth > self.healthPerBreed:
        if(len(self.grid[father.x][father.y])):
          creature.breed(self.grid[father.x][father.y][0])
    return

  def findClosestAllyAndBreed(self,creature):
    nearestCreature = self.findNearestFriendlyCreatureXY(creature) 
    if nearestCreature is not None: 
      path = self.findAPathD(creature.x, creature.y, nearestCreature.x, nearestCreature.y, creature.movementLeft)
      while len(path) > 0 and creature.movementLeft > 0 and self.distance(creature.x, creature.y, nearestCreature.x, nearestCreature.y) > 1 and creature.currentHealth > self.healthPerMove:
        creature.move(path[0][0], path[0][1])         
        path.remove(path[0])
      if self.distance(creature.x, creature.y, nearestCreature.x, nearestCreature.y) == 1 and creature.canBreed == True and creature.currentHealth > self.healthPerBreed:
        if(len(self.grid[nearestCreature.x][nearestCreature.y])):
          creature.breed(self.grid[nearestCreature.x][nearestCreature.y][0])
    return
  
  def findNearestEdiblePlantXY(self, creature, plantSize):
    #Doesn't work if there are no plants left to eat
    ediblePlants = [target for target in self.plants if target.size > plantSize]
    if len(ediblePlants) > 0:
      #Creates a dictionary of the distances to each plant that has a size > 0
      dict = {target.id:math.floor(self.distance(creature.x, creature.y, target.x, target.y)) for target in ediblePlants}       
      #Returns the object that is the closest creature  
      bestID = min(dict, key=dict.get)
      bestPlant = next(target for target in self.plants if target.id == bestID)
      #Returns the x and y value of that plant's location
      
      print "BEST Plant: " + str(bestPlant.x) + ", " + str(bestPlant.y)
      
      return bestPlant
    return None
  
  def findNearestEnemyAndEat(self, creature):
    nearestCreature = self.findNearestEnemyCreatureXY(creature) 
    if nearestCreature is not None:    
      path = self.findAPathD(creature.x, creature.y, nearestCreature.x, nearestCreature.y, creature.movementLeft)
      #Move until we are to the plant or can move no more
      while len(path) > 0 and creature.movementLeft > 0 and self.distance(creature.x, creature.y, nearestCreature.x, nearestCreature.y) > 1:
        creature.move(path[0][0], path[0][1])         
        path.remove(path[0])
      if self.distance(creature.x, creature.y, nearestCreature.x, nearestCreature.y) == 1 and creature.canEat == True:
        print "Attacking enemy"
        creature.eat(nearestCreature.x, nearestCreature.y) 
  
  def findAndEatPlant(self, creature):
    plant = self.findNearestEdiblePlantXY(creature, 3) 
    if plant is not None: 
      plant = self.findNearestEdiblePlantXY(creature, 0)
    if plant is not None:
      path = self.findAPathD(creature.x, creature.y, plant.x, plant.y, creature.movementLeft)
      #Move until we are to the plant or can move no more
      while len(path) > 0 and creature.movementLeft > 0 and self.distance(creature.x, creature.y, plant.x, plant.y) > 1:
        creature.move(path[0][0], path[0][1])         
        path.remove(path[0])
      #We are next to a plant! Eat it!
      if self.distance(creature.x, creature.y, plant.x, plant.y) == 1 and creature.canEat == True:
        creature.eat(plant.x, plant.y)
    return
  
  def findClosestHerbivoreAndBreed(self, creature):
    
    pass
  
  def herbAction(self, creature):
    x_mid = math.floor(self.mapWidth/2)
    y_mid = math.floor(self.mapHeight/2)
    if creature.currentHealth < math.floor((1.5*creature.maxHealth/3)):
      self.findAndEatPlant(creature)
    elif creature.currentHealth > self.healthPerBreed + 15:
      self.findClosestXAndBreed(creature, "herbivorism")
    if creature.movementLeft > 0:
      self.moveToMiddle(creature)
    return
  
  def carnAction(self, creature):
    x_mid = math.floor(self.mapWidth/2)
    y_mid = math.floor(self.mapHeight/2)
    
    
           
    self.moveToMiddle(creature)
    return
  
  ##This function is called each time it is your turn
  ##Return true to end your turn, return false to ask the server for updated information
  def run(self):
    self.grid = [[[] for _ in range(self.mapHeight)] for _ in range(self.mapWidth)] 
    self.statCreatureList = {"energy": [], "carnivorism": [], "herbivorism": [], "speed": [], "defense": []}  
    
    self.CHcreatureList = {"defense": [], "herbivorism": []}
    
    #fill grid with info     
    for plant in self.plants:
      self.grid[plant.x][plant.y] = [plant]
    for creature in self.creatures:
      self.grid[creature.x][creature.y] = [creature]  
      
      if creature.owner == self.playerID and creature.currentHealth > 0:
        stats = {"energy": creature.energy, "carnivorism": creature.carnivorism, "herbivorism": creature.herbivorism, "defense": creature.defense,"speed": creature.speed} 
        chstats = {"defense" : creature.defense, "herbivorism" : creature.herbivorism}
        self.statCreatureList[max(stats, key=stats.get)].append(creature)      
        self.CHcreatureList[max(chstats, key=chstats.get)].append(creature)
    
    myCreatures = [creature for creature in self.creatures if creature.owner == self.playerID]
        
    for creature in self.CHcreatureList["defense"]:
      self.carnAction(creature)
    
    for creature in self.CHcreatureList["herbivorism"]:
      self.herbAction(creature)
      
    print self.turnNumber
    return 1

  def __init__(self, conn):
    BaseAI.__init__(self, conn)
   

File: testClients/Tyler_cpp/AI.cpp
#include "AI.h"
#include "util.h"
#include <cstdlib>
#include <cmath>

AI::AI(Connection* conn) : BaseAI(conn) {}

const char* AI::username()
{
  return "darklordtyler";
}

const char* AI::password()
{
  return "password";
}
bool AI::isValid(int x, int y)
{
   bool valid = true;
   if(x >= mapWidth() || x < 0 || y >= mapHeight() || y < 0)
   {
      valid = false;
   }
   for(int i = 0; i < mappables.size(); i++)
   {
      if(mappables[i].x() == x && mappables[i].y() == y)
      {
         
         valid = false;
      }
   }
   return valid;
}
/*
float distance(float x1, float x2, float y1, float y2)
{
   return sqrt(pow(x1 - x2, 2) + pow(y1 - y2, 2));
}
*/
Creature* AI::findAdjacentEnemy(Creature* & c)
{
   int creaturex = c -> x();
   int creaturey = c -> y();
   Creature* tmp = NULL;
   
   for(int i = 0; i < creatures.size(); i++)
   {
      if(creatures[i].owner() != playerID())
      {
         if(creatures[i].x() == creaturex + 1 && creatures[i].y() == creaturey) // right
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex - 1 && creatures[i].y() == creaturey) // left
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey + 1) // down
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey - 1) // up
         {
            tmp = &creatures[i];
         }
      }
   }
   
   return tmp;
}

Creature* AI::findAdjacentFriend(Creature* & c)
{
   int creaturex = c -> x();
   int creaturey = c -> y();
   Creature* tmp = NULL;
   
   for(int i = 0; i < creatures.size(); i++)
   {
      if(creatures[i].owner() == playerID())
      {
         if(creatures[i].x() == creaturex + 1 && creatures[i].y() == creaturey) // right
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex - 1 && creatures[i].y() == creaturey) // left
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey + 1) // down
         {
            tmp = &creatures[i];
         }
         if(creatures[i].x() == creaturex && creatures[i].y() == creaturey - 1) // up
         {
            tmp = &creatures[i];
         }
      }
   }
   
   return tmp;
}

Plant* AI::findAdjacentPlant(Creature* & c)
{
   int creaturex = c -> x();
   int creaturey = c -> y();
   Plant* tmp = NULL;
   
   for(int i = 0; i < plants.size(); i++)
   {
      if(plants[i].x() == creaturex + 1 && plants[i].y() == creaturey) // right
      {
         tmp = &plants[i];
      }
      if(plants[i].x() == creaturex - 1 && plants[i].y() == creaturey) // left
      {
         tmp = &plants[i];
      }
      if(plants[i].x() == creaturex && plants[i].y() == creaturey + 1) // down
      {
         tmp = &plants[i];
      }
      if(plants[i].x() == creaturex && plants[i].y() == creaturey - 1) // up
      {
         tmp = &plants[i];
      }
   }
   
   return tmp;
}

bool AI::goLeft(Creature* & c)
{
   if(isValid(c -> x() - 1, c -> y()))
   {
      c -> move(c -> x() - 1, c -> y());
      return true;
   }
   return false;
}

bool AI::goRight(Creature* & c)
{
   if(isValid(c -> x() + 1, c -> y()))
   {
      c -> move(c -> x() + 1, c -> y());
      return true;
   }
   return false;
}

bool AI::goUp(Creature* & c)
{
   if(isValid(c -> x(), c -> y() - 1))
   {
      c -> move(c -> x(), c -> y() - 1);
      return true;
   }
   return false;
}

bool AI::goDown(Creature* & c)
{
   if(isValid(c -> x(), c -> y() + 1))
   {
      c -> move(c -> x(), c -> y() + 1);
      return true;
   }
   return false;
}

void AI::stupidMove(Creature* & c, int destx, int desty)
{
   int moves = 0;
   int max = (c -> movementLeft() / 2) + 1;
   bool right = true;
   bool up = false;
   bool left = false;
   bool down = false;
   int badrun = 2;
   int trial = 0;
   
   while(moves < max)
   {
      int x = c -> x();
      int y = c -> y();
      
      if((desty - y > destx - x && desty - y > 0 && trial == 0) || (badrun == 4 && trial != 0))
      {
         if(goDown(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 1;
            trial++;
         }
      }
      else if((destx - x > desty - y && destx - x > 0 && trial == 0) || (badrun == 1 && trial != 0))
      {
         if(goRight(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 2;
            trial++;
         }
      }
      else if((destx - x > desty - y && desty - y < 0 && trial == 0) || (badrun == 2 && trial != 0))
      {
         if(goUp(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 3;
            trial++;
         }
      }
      else if((((desty - y > destx - x && destx - x < 0) || destx - x == desty - y) && trial == 0) || (badrun == 3 && trial != 0))
      {
         if(goLeft(c))
         {
            moves++;
            trial = 0;
            badrun = 2;
         }
         else
         {
            badrun = 4;
            trial++;
         }
      }
      else
      {
         return;
      }
   }
}

int AI::findEnemyQuadrant()
{
   int first = 0;
   int second = 0;
   int third = 0;
   int fourth = 0;
   int max = 0;
   int quad = 0;
   int centerx = mapWidth() / 2;
   int centery = mapHeight() / 2;
   
   for(int i = 0; i < creatures.size(); i++)
   {
      if(creatures[i].owner() != playerID())
      {
         if(creatures[i].x() > centerx && creatures[i].y() < centery)
         {
            first++;
            if(first > max)
            {
               max = first;
               quad = 1;
            }
         }
         else if(creatures[i].x() < centerx && creatures[i].y() < centery)
         {
            second++;
            if(second > max)
            {
               max = second;
               quad = 2;
            }
         }
         else if(creatures[i].x() < centerx && creatures[i].y() > centery)
         {
            third++;
            if(third > max)
            {
               max = third;
               quad = 3;
            }
         }
         else
         {
            fourth++;
            if(fourth > max)
            {
               max = fourth;
               quad = 4;
            }
         }
      }
   }
   return quad;
}

int AI::findPlantQuadrant()
{
   int first = 0;
   int second = 0;
   int third = 0;
   int fourth = 0;
   int max = 0;
   int quad = 0;
   int centerx = mapWidth() / 2;
   int centery = mapHeight() / 2;
   
   for(int i = 0; i < plants.size(); i++)
   {
      if(plants[i].x() > centerx && plants[i].y() < centery)
      {
         first++;
         if(first > max)
         {
            max = first;
            quad = 1;
         }
      }
      else if(plants[i].x() < centerx && plants[i].y() < centery)
      {
         second++;
         if(second > max)
         {
            max = second;
            quad = 2;
         }
      }
      else if(plants[i].x() < centerx && plants[i].y() > centery)
      {
         third++;
         if(third > max)
         {
            max = third;
            quad = 3;
         }
      }
      else
      {
         fourth++;
         if(fourth > max)
         {
            max = fourth;
            quad = 4;
         }
      }
   }
   return quad;
}

//This function is run once, before your first turn.
void AI::init(){}


//This function is called each time it is your turn.
//Return true to end your turn, return false to ask the server for updated information.
bool AI::run()
{
  Creature* myCreatures[100];
  Creature* enemies[100];
  Creature* bestCreatures[5];
  
  
  int numMyCreatures = 0;
  int numEnemies = 0;
  int position1 = 0;
  int position2 = 0;
  
  
  // dudes with max stats
  int maxStrength = 0;
  int maxHerbivore = 0;
  int maxCarnivore = 0;
  int maxSpeedster = 0;
  int maxShield = 0;
  
  
  int centerx = mapWidth() / 2;
  int centery = mapHeight() / 2;
  int plantx = 0;
  int planty = 0;
  float dist = mapWidth();
  
  int pquadrant = findPlantQuadrant();
  int equadrant = findEnemyQuadrant();
  int enemiesx = 0;
  int enemiesy = 0;
  
   
  for(int i = 0; i < creatures.size(); i++)
  {
    if(creatures[i].owner() == playerID())
    {
      myCreatures[position1] = &creatures[i];
      numMyCreatures++;
      position1++;
      int health = creatures[i].maxHealth();
      int herb = creatures[i].herbivorism();
      int carn = creatures[i].carnivorism();
      int speed = creatures[i].speed();
      int def = creatures[i].defense();
      if(health > maxStrength)
      {
         bestCreatures[0] = &creatures[i];
      }
      if(herb > maxHerbivore)
      {
         bestCreatures[1] = &creatures[i];
      }
      if(carn > maxCarnivore)
      {
         bestCreatures[2] = &creatures[i];
      }
      if(speed > maxSpeedster)
      {
         bestCreatures[3] = &creatures[i];
      }
      if(def > maxShield)
      {
         bestCreatures[4] = &creatures[i];
      }
    }
    else
    {
      enemies[position2] = &creatures[i];
      numEnemies++;
      position2++;
    }
  }
  /*
  switch(equadrant)
  {
    case 1:
      enemiesx = centerx + (centerx / 2);
      enemiesy = centery / 2;
      break;
    case 2:
      enemiesx = centerx / 2;
      enemiesy = centery / 2;
      break;
    case 3:
      enemiesx = centerx / 2;
      enemiesy = centery + (centery / 2);
      break;
    case 4:
      enemiesx = centerx + (centerx / 2);
      enemiesy = centery + (centery / 2);
      break;
  }
  */
  for(int i = 0; i < numEnemies; i++)
  {
    int enx = enemies[i] -> x();
    int eny = enemies[i] -> y();
    if(equadrant == 1)
    {
      if(enx >= centerx && eny <= centery)
      {
         enemiesx = enx;
         enemiesy = eny;
      }
    }
    else if(equadrant == 2)
    {
      if(enx <= centerx && eny <= centery)
      {
         enemiesx = enx;
         enemiesy = eny;
      }
    }
    else if(equadrant == 3)
    {
      if(enx <= centerx && eny >= centery)
      {
         enemiesx = enx;
         enemiesy = eny;
      }
    }
    else if(equadrant == 4)
    {
      if(enx >= centerx && eny >= centery)
      {
         enemiesx = enx;
         enemiesy = eny;
      }
    }
  }
  
  
  
  switch(pquadrant)
  {
    case 1:
      plantx = centerx + (centerx / 2);
      planty = centery / 2;
      break;
    case 2:
      plantx = centerx / 2;
      planty = centery / 2;
      break;
    case 3:
      plantx = centerx / 2;
      planty = centery + (centery / 2);
      break;
    case 4:
      plantx = centerx + (centerx / 2);
      planty = centery + (centery / 2);
      break;
  }
  
  for(int i = 0; i < numMyCreatures; i++)
  {
    int movesLeft = myCreatures[i] -> movementLeft();
    int xloc = myCreatures[i] -> x();
    int yloc = myCreatures[i] -> y();
    int health = myCreatures[i] -> maxHealth();
    int herb = myCreatures[i] -> herbivorism();
    int carn = myCreatures[i] -> carnivorism();
    int speed = myCreatures[i] -> speed();
    int def = myCreatures[i] -> defense();
    Creature* food = NULL;
    Creature* mate = NULL;
    Plant* shrub = NULL;
    
    if(myCreatures[i] -> canEat())
    {
      food = findAdjacentEnemy(myCreatures[i]);
      shrub = findAdjacentPlant(myCreatures[i]);
      if(food != NULL)
      {
         myCreatures[i] -> eat(food -> x(), food -> y());
      }
      else if(shrub != NULL)
      {
         myCreatures[i] -> eat(shrub -> x(), shrub -> y());
      }
    }
    
    if(carn > 3)
      stupidMove(myCreatures[i], enemiesx, enemiesy);
    else if(herb > 3)
      stupidMove(myCreatures[i], plantx, planty);
    else
      stupidMove(myCreatures[i], centerx, centery);
    
    if(myCreatures[i] -> canBreed())
    {
      mate = findAdjacentFriend(myCreatures[i]);
      if(mate != NULL)
      {
         myCreatures[i] -> breed(*mate);
         stupidMove(myCreatures[i], centerx, centery);
      }
    }
    
    /*
    if(carn > 3)
      stupidMove(myCreatures[i], enemiesx, enemiesy);
    if(herb > 3)
      stupidMove(myCreatures[i], plantx, planty);
    if(herb <= 3 && carn <= 3)
      stupidMove(myCreatures[i], centerx, centery);
    //stupidMove(myCreatures[i], enemiesx, enemiesy);
    */
   
  }
  
  return true;
}

//This function is run once, after your last turn.
void AI::end(){}
